### DISCUSSION TITLE
Validate XSDs with external references

### DISCUSSION THREAD

**Discussion by DmitryMaletin** [Apicurio/apicurio-registry#5051] [Q&A]
Is it possible somehow to upload and validate XSDs with external references?

**Comment by EricWittmann** [comment]
It's certainly possible to upload XSDs that have references to other schemas.  However, we haven't really implemented the same level for support as we have with other types like Avro.  Registry doesn't have a ton of legacy web services users, so our maturity for XSD and WSDL is lower.

We'd love to collaborate in this area though if you were willing.  I would be happy to guide you on what you would need to do to contribute!

**Reply by EricWittmann** [reply]
BTW what I mean is that we simply haven't implemented some of the components needed to properly implement all of the Rules for the XSD type.  Specifically here:

https://github.com/Apicurio/apicurio-registry/blob/main/schema-util/util-provider/src/main/java/io/apicurio/registry/types/provider/XsdArtifactTypeUtilProvider.java

You can see that we have real implementations of Validity and Canonicalization.  But we don't have any implementation for:

* `ContentDereferencer`
* `ReferenceFinder`

We would need to implement those two interfaces for XSD files.

**Reply by DmitryMaletin** [reply]
Thanks for response. 
XSD is still quite heavily used in finance services and that probably will remain for a long  time. So I see it quite beneficial what apicurio registry can also host them. Sadly for now I have to switch off validation to deploy them and that a little bit incontinent. 
So I would be glad to get any additional guidance to understand how to add that support.

**Reply by EricWittmann** [reply]
OK great!  I think the first step would be to create a test: put together an XSD that uses types from another schema (or better, two other schemas).  Then we can work on how those should be modeled in Registry (I have some ideas on this).

Once that is done we can create various JUnit tests and then implement the code.  We will definitely need to implement the following interfaces for XSD files:

* `ContentDereferencer` : https://github.com/Apicurio/apicurio-registry/blob/main/schema-util/common/src/main/java/io/apicurio/registry/content/dereference/ContentDereferencer.java
* `ReferenceFinder` : https://github.com/Apicurio/apicurio-registry/blob/main/schema-util/common/src/main/java/io/apicurio/registry/content/refs/ReferenceFinder.java

We'll also need to enhance the https://github.com/Apicurio/apicurio-registry/blob/main/schema-util/xsd/src/main/java/io/apicurio/registry/rules/validity/XsdContentValidator.java class to do a better job of validation and validation of references.

I have some good ideas on how this all should probably work.
